lib/compat/wordpress-6.1/class-wp-theme-json-6-1.php

```diff
@@ -462,6 +496,27 @@ public function get_styles_for_block( $block_metadata ) {
 		$selector = $block_metadata['selector'];
 		$settings = _wp_array_get( $this->theme_json, array( 'settings' ) );
 
+		// Process style declarations for block support features the current
+		// block contains selectors for. Values for a feature with a custom
+		// selector are filtered from the theme.json node before it is
+		// processed as normal.
+		$feature_declarations = array();
+
+		if ( ! empty( $block_metadata['features'] ) ) {
+			foreach ( $block_metadata['features'] as $feature_name => $feature_selector ) {
+				if ( ! empty( $node[ $feature_name ] ) ) {
+					// Create temporary node containing only the feature data
+					// to leverage existing `compute_style_properties` function.
+					$feature = array( $feature_name => $node[ $feature_name ] );
+					// Generate the feature's declarations only.
+					$feature_declarations[ $feature_selector ] = static::compute_style_properties( $feature, $settings, null, $this->theme_json );
```

Comment:
I'm not sure what I'm seeing is correct, but I can't get multiple features to work at the same time, and I suspect it's because the features are either being unset or overwritten here.

So the border styles won't show when I have color/spacing/typography defined. 

<details>

<summary>block.json supports example</summary>

```json
	"supports": {
		"anchor": true,
		"align": true,
		"color": {
			"__experimentalSkipSerialization": true,
			"gradients": true,
			"__experimentalDefaultControls": {
				"background": true,
				"text": true
			},
			"__experimentalSelector": "> table"
		},
		"typography": {
			"fontSize": true,
			"lineHeight": true,
			"__experimentalFontStyle": true,
			"__experimentalFontWeight": true,
			"__experimentalLetterSpacing": true,
			"__experimentalTextTransform": true,
			"__experimentalDefaultControls": {
				"fontSize": true
			},
			"__experimentalSelector": "> table"
		},
		"__experimentalBorder": {
			"__experimentalSkipSerialization": true,
			"color": true,
			"style": true,
			"width": true,
			"__experimentalSelector": "> table",
			"__experimentalDefaultControls": {
				"color": true,
				"style": true,
				"width": true
			}
		},
		"__experimentalSelector": ".wp-block-table",
		"__experimentalStyle": {
			"spacing": {
				"margin": "0 0 1em 0"
			}
		}
	},

```

</details>

<details>

<summary>theme.json styles</summary>

```json
"styles": {
		"blocks": {
			"core/cover": {
				"border": {
					"color": "fuchsia",
					"style": "dashed",
					"width": "3px"
				},
				"color": {
					"text": "fuchsia",
					"background": "black"
				},
				"typography": {
					"fontSize": "1px"
				},
				"spacing": {
					"padding": "100px"
				}
			},
			"core/table": {
				"border": {
					"color": "fuchsia",
					"style": "dashed",
					"width": "3px"
				},
				"color": {
					"text": "fuchsia",
					"background": "black"
				},
				"typography": {
					"fontSize": "100px"
				},
				"spacing": {
					"padding": "100px"
				}
			}
		}
	},

```

</details>

The only way I can get it to work is:

```php
					// Generate the feature's declarations only.
					if ( isset( $feature_declarations[ $feature_selector ] ) ) {
						$feature_declarations[ $feature_selector ] = array_merge( $feature_declarations[ $feature_selector ], static::compute_style_properties( $feature, $settings, null, $this->theme_json ) );
					} else {
						$feature_declarations[ $feature_selector ] = static::compute_style_properties( $feature, $settings, null, $this->theme_json );
					}
```

But it might an indicator of a deeper issue. 

It could be related to the fact that I'm seeing duplicate rules in the HTML source? Here's what I'm seeing when testing the cover block:

```html

<style id='global-styles-inline-css'> \* lots of styles here*\ .wp-block-cover > .wp-block-cover__inner-container > p{padding: 100px;} \* lots of styles here*\ .
</style>


<style id='wp-block-cover-inline-css'>
    .wp-block-cover{font-size: 1px;}.wp-block-cover > .wp-block-cover__inner-container > p{padding: 100px;}
</style>

```

Reply:
Thanks for catching this issue @ramonjd. I'd clearly been caught up testing different supports targeting different inner elements.

I've fixed this by [leveraging your snippet](https://github.com/WordPress/gutenberg/pull/42087/commits/e02ea3591df2b01ac6d048690b66428e680aa354#diff-f28aa79997745a4a51f917012b45a809c9b45d55134b55379a718f5c751dddf4R517-R521), adding an [explanatory inline comment](https://github.com/WordPress/gutenberg/pull/42087/commits/e02ea3591df2b01ac6d048690b66428e680aa354#diff-f28aa79997745a4a51f917012b45a809c9b45d55134b55379a718f5c751dddf4R514-R516), and [expanding the unit tests](https://github.com/WordPress/gutenberg/pull/42087/commits/e02ea3591df2b01ac6d048690b66428e680aa354#diff-a51f75ea7a86ddd18edea3fe94a0482427795fcd7b247e6959149cbac60964dfR108-R120) to cover:

- two supports using a shared custom feature level selector e.g. `.inner`
- one other support using its own unique selector e.g. `.sub-heading`
- and, a further support that doesn't have a feature level selector specified.


> But it might an indicator of a deeper issue.
> 
> It could be related to the fact that I'm seeing duplicate rules in the HTML source? Here's what I'm seeing when testing the cover block:

This looks to be unrelated to me. There was definitely a flaw in the previous logic generating the feature styles.

Reply:
> It could be related to the fact that I'm seeing duplicate rules in the HTML source? Here's what I'm seeing when testing the cover block:

Thanks for looking into it. 

I'm also seeing it on trunk and suspect it's because we're enqueuing block styles in both places after https://github.com/WordPress/gutenberg/pull/42005